[ITBL-5309] Add handleUniversalLink: method

vbabenkoru opened the pull request.

vbabenkoru committed:
Add resolveAppLinkURL:callback: method with a URL callback in addition to getAndTrackDeeplink

vbabenkoru committed:
Make sure we call getAndTrackDeeplink: and resolveApplinkURL: on the main queue

vbabenkoru committed:
Add tests for resolveApplinkURL: method, check that the callback is run on the main thread

tapashmajumder commented on Iterable-iOS-SDK/IterableAPI.h:
```diff
@@ -513,7 +513,16 @@ typedef void (^OnFailureHandler)(NSString *reason, NSData *_Nullable data);
  @discussion            passes the string of the redirected URL to the callback, returns the original webpageURL if not an iterable link
  */
-+(void) getAndTrackDeeplink:(NSURL *)webpageURL callbackBlock:(ITEActionBlock)callbackBlock;
++ (void)getAndTrackDeeplink:(NSURL *)webpageURL callbackBlock:(ITEActionBlock)callbackBlock;
```
Do we want to deprecate this call?

tapashmajumder reviewed (approved):
+1

davidtruong commented on Iterable-iOS-SDK/IterableAPI.m:
```diff
@@ -457,12 +457,19 @@ + (IterableAPI *)sharedInstanceWithApiKey:(NSString *)apiKey andUserId:(NSString
     [deeplinkManager getAndTrackDeeplink:webpageURL callbackBlock:callbackBlock];
 }
 
++ (void)resolveApplinkURL:(NSURL *)url callback:(ITBURLCallback)callback;
```
resolveAppLink sounds like a function that should automatically handle calling openUrl. 
Do you think we should do that automatically?

We can overload getAndTrackDeeplink to pass back a URL.

tapashmajumder replied:
I think we should let the app handle logic for handling the url opens. This is how we have it in the sample app
```
        IterableAPI.resolve(applinkURL: url) { (resolvedUrl) in
            if let resolvedUrl = resolvedUrl {
                DeeplinkHandler.handle(url: resolvedUrl) // DeeplinkHandler is their class.
            }
        }

```

andyrova replied:
Making the example `DeeplinkHandler` code available either in support docs or a repository would be helpful for clients

davidtruong reviewed (commented):
Should we automatically handle the redirects and open the url?

vbabenkoru committed:
Replace resolveApplinkURL: with handleUniversalLink: that creates a new IterableAction and executes it with IterableActionRunner
Add IterableActionContext specifying the context the action is executing in (including the source – push or universal link)

vbabenkoru committed:
Fix warnings

tapashmajumder commented on Iterable-iOS-SDK/IterableAPI.h:
```diff
@@ -513,7 +513,18 @@ typedef void (^OnFailureHandler)(NSString *reason, NSData *_Nullable data);
+ * Handles a Universal Link
+ * For Iterable links, it will track the click and retrieve the original URL,
+ * pass it to `IterableURLDelegate` for handling
+ * If it's not an Iterable link, it just passes the same URL `IterableURLDelegate`
```
passes the same URL *to* `IterableURLDelegate`

vbabenkoru committed:
Fix a typo

vbabenkoru committed:
Fix another typo

davidtruong reviewed (approved):
LGTM
